gunrock/oprtr/intersection/kernel.cuh

```diff
@@ -179,21 +178,65 @@ cudaError_t Launch(const GraphT &graph, const FrontierInT *frontier_in,
   typedef typename Dispatch<FLAG, InKeyT, OutKeyT, SizeT, ValueT, VertexT,
                             InterOpT, true>::KernelPolicyT KernelPolicyT;
 
+  util::SaveToRestore state;
+  state.Save();
+
   size_t input_length = graph.edges;
   size_t stride =
       (input_length + KernelPolicyT::BLOCKS * KernelPolicyT::THREADS - 1) >>
       (KernelPolicyT::LOG_THREADS + KernelPolicyT::LOG_BLOCKS);
-  SizeT num_vertex = graph.nodes;
-  SizeT num_edges = graph.edges;
-
-  IntersectTwoSmallNL<FLAG, InKeyT, OutKeyT, SizeT, ValueT, VertexT, InterOpT>
-      <<<KernelPolicyT::BLOCKS, KernelPolicyT::THREADS, 0, parameters.stream>>>(
-          graph.CsrT::row_offsets.GetPointer(util::DEVICE),
-          graph.CsrT::column_indices.GetPointer(util::DEVICE),
-          frontier_in->GetPointer(util::DEVICE),
-          graph.CsrT::column_indices.GetPointer(util::DEVICE),
-          frontier_out->GetPointer(util::DEVICE), input_length, stride,
-          num_vertex, num_edges, inter_op);
+
+  SizeT num_vertices = graph.nodes;
+
+  SizeT edges_per_gpu = gunrock::util::ceil_divide(input_length, mgpu_context.getGpuCount());
+
+  printf("Num verts %d\n", num_vertices);
+  printf("edges_per_gpu %d\n", edges_per_gpu);
+
+  std::vector<std::thread> threads;
+  threads.reserve(mgpu_context.getGpuCount());
+
+  for (int i = 0; i < mgpu_context.getGpuCount(); ++i) {
+    auto context = mgpu_context.contexts[i];
+    auto rows_offset = edges_per_gpu * i + graph.CsrT::row_offsets.GetPointer(util::DEVICE);
+    auto cols_offset = edges_per_gpu * i + graph.CsrT::column_indices.GetPointer(util::DEVICE);
+
+    printf("Rows offset %p\n", rows_offset);
+    printf("Cols offset %p\n", cols_offset);
+
+    threads.push_back(std::thread( [&, context]() {
+        
+        IntersectTwoSmallNL<FLAG, InKeyT, OutKeyT, SizeT, ValueT, VertexT, InterOpT>
+        <<<KernelPolicyT::BLOCKS, KernelPolicyT::THREADS, 0, parameters.stream>>>(
+            rows_offset,
+            cols_offset,
+            frontier_in->GetPointer(util::DEVICE),
+            cols_offset,
+            frontier_out->GetPointer(util::DEVICE),
+            edges_per_gpu,
+            stride,
+            inter_op);
+    }));
+  }
```

Comment:
I'm not sure about parts of this selection and this bit ` threads.push_back(std::thread( [&, context]()` . The reference in the lambda capture could work, but makes me think a number of the variables could change before the kernel is actually called.

I recommend using a range-based for loop over the `mgpu_context.contexts`and  moving the `rows_offsets` and `cols_offsets` inside of the `threads` lambda function. You can then use the `context.device_id` to compute your offsets (the `device_id` is guaranteed to match / have the same range of values as your `i` in the currently implemented code).

Reply:
Thanks this is helpful.